### ISSUE TITLE
custom objective function using features 

### ISSUE THREAD

**Issue by hep07** [dmlc/xgboost#3364]
I am working with a very special model where the loss function is not standard. Most importantly it depends on certain features in X, not just predictions and labels.  Following previous issues I under stand that I could define grad and hess of my obj the following way, 
```

# user define objective function, given prediction, return gradient and second order gradient
def logregobj( preds, dtrain ):
    labels = dtrain.get_label()
    grad = preds - labels
    hess = preds * (1.0-preds)
    return grad, hess

```
The problem is that my loss function depend not only on preds and labels, but also on certain features in dtrain. Does the second argument in user defined obj have to be Dmatrix type?  I figured from previous issues that we weren't able to extract feature values from dtrain. Is that still the case, even if I only need one column in the feature matrix, which has the same dimension as label? If Yes, is there any work around so that I could code my loss functions? 

Edit: 
With more digging into the repo, I found something in the R-package/demo/custom_objective.R
```
# there can be cases where you want additional information 
# being considered besides the property of DMatrix you can get by getinfo
# you can set additional information as attributes if DMatrix

# set label attribute of dtrain to be label, we use label as an example, it can be anything 
attr(dtrain, 'label') <- getinfo(dtrain, 'label')
# this is new customized objective, where you can access things you set
# same thing applies to customized evaluation function
logregobjattr <- function(preds, dtrain) {
  # now you can access the attribute in customized function
  labels <- attr(dtrain, 'label')
  preds <- 1/(1 + exp(-preds))
  grad <- preds - labels
  hess <- preds * (1 - preds)
  return(list(grad = grad, hess = hess))
}
```
It is sort of what I need but my follow-up question is: 

just use label as an example, I am not sure how this objective function will be called. If there are some subset operation to the DMat class, say take only the first 50% rows from the training data, then the attributes, treated as meta-data will not be proper subsetted and as a result the obj function will not perform as expected. 

Pu

**Comment by tqchen** [comment]
questions are now encouraged at https://discuss.xgboost.ai/

**Comment by tqchen** [comment]
I think you find the right answer of the customized loss
